c-bindings/src/data_serializable_raw.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iterator>
+
+// C++ client public headers
+#include "geode/DataInput.hpp"
+#include "geode/DataOutput.hpp"
+#include "geode/internal/DataSerializablePrimitive.hpp"
+
+// C client public headers
+#include "data_serializable_raw.hpp"
+
+// C client private headers
+
+using apache::geode::client::DataInput;
+using apache::geode::client::DataOutput;
+
+namespace apache {
+namespace geode {
+namespace client {
+namespace internal {
+
+DataSerializableRaw::DataSerializableRaw(const int8_t* data, size_t size) {
+  bytes_.reserve(size);
+  std::copy(data, data + size, std::back_inserter(bytes_));
+}
+
+std::shared_ptr<DataSerializableRaw> DataSerializableRaw::create(
+    const int8_t* data, size_t size) {
+  return std::make_shared<DataSerializableRaw>(data, size);
+}
+
+void DataSerializableRaw::toData(DataOutput& dataOutput) const {
+  dataOutput.writeBytesOnly(bytes_.data() + dsCodeSize_,
+                            bytes_.size() - dsCodeSize_);
+}
+
+void DataSerializableRaw::fromData(DataInput& dataInput) {
+  dataInput.readBytesOnly(bytes_.data() + dsCodeSize_,
+                          bytes_.size() - dsCodeSize_);
+}
+
+DSCode DataSerializableRaw::getDsCode() const {
+  return static_cast<DSCode>(bytes_[0]);
+}
+
+bool DataSerializableRaw::operator==(const CacheableKey& other) const {
+  if (auto otherKey = dynamic_cast<const DataSerializableRaw*>(&other)) {
+    return bytes_ == otherKey->bytes_;
+  }
+
+  return false;
+}
+
+int32_t DataSerializableRaw::hashcode() const {
+  if (hashCode_ == 0) {
+    hashCode_ = internal::geode_hash<std::vector<int8_t>>{}(bytes_);
```

Comment:
This is going to result in a hash on the client that is not consistent with the hash on the server. The hash value needs to come from the pre-serialized form of the object being serialized here. This will result in sub-optimal put operations.

Reply:
Good catch! I think we're going to have to create the CacheableKey objects at the C_Bindings layer, similar to what clicache does.

Reply:
The probably with `CacheableKey` is that it expects an equality rule that isn't necessary at the serialization layer but is at the local cache layer. The C-binding is going to need to provide a function pointer, or something I can't think of right now, to solve that local caching key issue. Value objects don't need hash or equality.

Reply:
When I thought about this problem years ago I felt that the C-binding layer would either need to below the local caching and serialization layers and effectively be a networking layer, or the C-binding will need to keep track of pointers back to the original C types and some function pointers. Neither of them are going to be easy.

```
class VoidCachableKey {
  public void VoidCacheableKey(const void* key, int32_t (*hash)(const void *), bool (*equals)(const void *, const void *, const char* serialized, size_t serializedLength) : ... {}
  int32_t hashcode() {
    return hash_(key);
  }
  bool equals(const VoidCachableKey& other) {
    return equals_(key, other.key_);
  }
}
```

Reply:
@pivotal-jbarrett I think I need the short course in what we're using `hashcode` for that causes issues here.  I wasn't aware that we were comparing or otherwise using server-side hashes against client-side, which... sounds like a bad idea to begin with, but I have to admit I'm not surprised by.

Reply:
The hash is used to route the keys to the correct server. If the client and server don't agree on the hashing algorithm and result then they will silently work but it results in suboptimal operations with multiple hops.
Equals is only going to be used client side for local caching, if caching is enabled. Hash maps require both a hash and equality function. The equality on the client and server don't have to match but given the hash needs to match it silly not to have the equality match. One problem that could exist if you don't use constant equality is that the server and client could disagree on what keys are in the caches. Imagine an object that has a field that is serialized, like timestamp, that isn't part of the equality or hash because its ephemeral but we hash the serialized form then the server and client won't agree on the hash value. Then if caching was enabled and equality is also based on the serialized form then the sever would only have one Key('foo') with some timestamp value X and the client could have Key('foo') with timestamp Y and Key('foo') with timestamp X. Local cache lookups would fail as timestamp advances, which could cause massive heap exhaustion.

Reply:
Okay, so the `DataSerializableRaw` hack is essentially useless, is what I hear you saying, because without a way to implement  `hashcode` and `==` we can only use it for values.  Probably time for a new plan for keys.

Reply:
Yes, and for values you don't need the hash.